src/builtins.rs

```diff
@@ -257,9 +258,142 @@ lazy_static! {
                 }
             }
         ),
+        (
+            "LOWER_BOUND",
+            BuiltIn {
+                decl: "FUNCTION LOWER_BOUND<U: ANY> : DINT
```

Comment:
I understand why we need an ANY-generic here, but I wonder how we can validate if the user really passed an array here :-/

so can we validate if `LOWER_BOUND(x, 1)` and x is no array?

Reply:
It is possible to validate this in the builtin-annotation. However, this annotation is returned in the resolver stage and currently just unwrapped, so if we find an error there, it will panic. We also do not have a diagnostician in the resolver and I'm not sure if I should add one (or if it is even feasible to do so).

This is the way I have tried it:

```
fn validate_vla_builtins(
    annotator: &mut TypeAnnotator,
    operator: &AstStatement,
    parameters: Option<&AstStatement>,
    is_lower: bool,
) -> Result<(), Diagnostic> {
    let pou_name = if is_lower { "LOWER_BOUND" } else { "UPPER_BOUND" };

    let params = parameters.ok_or_else(|| {
        Diagnostic::codegen_error(&format!("{} requires parameters", pou_name), operator.get_location())
    })?;
    // Get the input and annotate it with a vla type
    let params = flatten_expression_list(params);
    let Some(input) = params.get(0) else {
            unreachable!()
    };
    let type_information =
        annotator.annotation_map.get_type_or_void(input, annotator.index).get_type_information();

    if !matches!(
        type_information,
        DataTypeInformation::Struct {
            source: StructSource::Internal(InternalType::VariableLengthArray { .. }),
            ..
        } | DataTypeInformation::Array { .. },
    ) {
        return Err(Diagnostic::SyntaxError {
            message: format!(
                "Invalid parameter type for {} function. Expected array or variable length array, found {}",
                pou_name,
                type_information.get_name()
            ),
            range: vec![operator.get_location()],
            err_no: ErrNo::call__invalid_parameter_type,
        });
    };

    Ok(())
}
```

Reply:
Does it make sense to add a diagnostician to builtins.rs and have it accept diagnostics?

Reply:
I have added an internal TypeNature for VLAs to solve this issue. In doing so I have noticed that arrays have type-nature ANY, which means they pass any check that checks for derives (derives_from any => true). I have not extensively checked if this causes other issues and it might be an isolated thing for builtins.